merge redesign

rafmudaf opened the pull request:
DO NOT MERGE! I'm opening this pull request so we can compare the differences since there are many. Comment here with anything that looks off and I'll fix it locally. Then, I'll push this branch directly to rhammond2/enhancement/wind_rose.

rafmudaf committed:
Update the example input file structure

rafmudaf committed:
Enable example 00

rafmudaf committed:
Add infrastructure for performance profiling

rafmudaf committed:
Prefer . over ! in logs

rafmudaf committed:
Remove FI reinitialize routine

FLORIS v3 does not require reinit. Instead, we create new FLORIS objects as needed.

rafmudaf committed:
Merge branch 'redesign' into v3/examples

rafmudaf committed:
Add model_string abstract property for wake models

rafmudaf committed:
Remove unused functionity in Farm

wtg-id
XArray

These should be added again when needed

rafmudaf committed:
Add profiling scripts

rafmudaf committed:
Remove the unused FarmController class

rafmudaf committed:
Remove convert_to_Vec3 utility

Just call Vec3 directly instead

rafmudaf committed:
Update tests for the single-turbine definition

rafmudaf committed:
Remove attrs from Floris and simplify init

rafmudaf committed:
Clean up: moving things around

rafmudaf committed:
Add a reference hub height property on Farm

rafmudaf committed:
Explicitly support a single turbine definition

This commit removes a lot of functionality from Farm and the
Turbine module that would enable supporting heterogeneous turbines.

rafmudaf committed:
Use scalars for elementwise multiplication

Instead of creating an array with constant values

rafmudaf committed:
Remove unused methods on the Grid

rafmudaf committed:
Reenable the full flow field solve

Temporarily removed the WAT calculation

rafmudaf committed:
Update Example 00

rafmudaf committed:
Include the coordinate rotation function

rafmudaf committed:
Update and consolidate example input files

rafmudaf committed:
Dont run profiling scripts in automated tests

rafmudaf committed:
Remove multi turbine support in test input files

rafmudaf committed:
Add dimension timing test

rafmudaf committed:
Bug fix: include deflection field in jensen model

rafmudaf committed:
Add TSR to turbine class

rafmudaf committed:
Update CutPlane visualization routines for v3

rafmudaf committed:
Update example 01

RHammond2 committed:
add error for missing parameters in FromDictMixin and update tests

RHammond2 committed:
change wind speed arrays to non-initialized values

RHammond2 committed:
make truth checking straightforward

rafmudaf committed:
Merge pull request #49 from RHammond2/v3/raf-performance

Better error messaging for FromDictMixin

rafmudaf committed:
Merge pull request #48 from rafmudaf/v3/performance_usability

General improvements for performance and usability

rafmudaf committed:
Merge branch 'redesign' into enhancement/wind_rose

# Conflicts:
#	examples/_getting_started/example_00_open_and_vis_floris.py
#	examples/_getting_started/example_01_basic_adjustments.py
#	examples/inputs/jensen.json
#	examples/other_jsons/input_legacy.json
#	src/floris/simulation/farm.py
#	src/floris/simulation/wake_deflection/gauss.py
#	src/floris/simulation/wake_deflection/jimenez.py
#	src/floris/simulation/wake_velocity/gauss.py
#	src/floris/simulation/wake_velocity/jensen.py
#	src/floris/tools/__init__.py
#	src/floris/tools/floris_interface.py
#	src/floris/tools/interface_utilities.py
#	src/floris/tools/visualization.py
#	src/floris/utilities.py

RHammond2 commented on examples/_getting_started/example_00_open_and_vis_floris.py:
```diff
@@ -42,16 +42,16 @@
 # the underlying classes
 fi = FlorisInterface(INPUT_JSON)
-fi.floris.farm.farm_controller.set_yaw_angles(np.array([25.0]))
+
+yaw_angles = np.zeros((1,1,3))
+yaw_angles[:,:,0] = 25.0
+fi.floris.farm.yaw_angles = yaw_angles
```
Either of these will work now that I've updated the control flow in the `set_yaw_angles` method, but it is best practice (I think) to match your shapes correctly for the examples.

RHammond2 commented on examples/_getting_started/example_01_basic_adjustments.py:
```diff
@@ -67,50 +65,67 @@ def plot_slice_shortcut(fi, ax, title):
+
 # Plot the initial setup
-fi.calculate_wake()
+fi.floris.solve_for_viz()
```
Is this going to be the final method name? I know the naming convention is a can we've been kicking down the road here.

rafmudaf replied:
It probably won't be the final name, but we will need to differentiate the visualization solver (or more accurately "full flow field" solver) from the other one. Maybe those are good names:
1. turbine grid solver
2. flow field grid solver

The differentiating features are the grid type.

RHammond2 commented on examples/_getting_started/example_01_basic_adjustments.py:
```diff
@@ -67,50 +65,67 @@ def plot_slice_shortcut(fi, ax, title):
 # Change the wind speed
-fi.reinitialize_flow_field(wind_speed=7.0)
-fi.calculate_wake()
+base_input_dict["flow_field"]["wind_speeds"] = [7.5]
+fi = FlorisInterface(base_input_dict)
+fi.floris.solve_for_viz()
```
The next handful of changes look like what we were talking about earlier with respect to how we intialize and change values. It could be good to show both as a means for demonstrating that v3 respects how we did things in v2 (for the most part, but that your change is the preferred method.

RHammond2 commented on examples/_getting_started/example_01_basic_adjustments.py:
```diff
@@ -67,50 +65,67 @@ def plot_slice_shortcut(fi, ax, title):
-plot_slice_shortcut(fi, axarr[8], "Change yaw angles")
-plot_turbines_with_fi(axarr[8], fi)
-
+# fi = FlorisInterface(base_input_dict)
```
I assume I'll get to this, but what is the reason for not adding in the yaw angles?

RHammond2 commented on examples/inputs/input_full_v3.yaml:
```diff
@@ -18,9 +18,7 @@ farm:
-  wtg_id:
-  - WTG-1
-  
+
```
Should we keep the wtg_id in here as an example?

RHammond2 commented on src/floris/simulation/base.py:
```diff
@@ -61,6 +61,15 @@ class BaseModel(BaseClass, ABC):
     create a valid model.
     """
 
+    @property
+    def model_string(self):
+        return self.model_string
+
+    @model_string.setter
+    @abstractmethod
+    def model_string(self, string):
+        raise NotImplementedError("BaseModel.model_string")
+
```
I'm not sure I'm following this change.

rafmudaf replied:
It forces the subclass to have this `@property`. This is the same as an abstract method but since its a `@property` is like an abstract attribute.

RHammond2 commented on src/floris/simulation/farm.py:
```diff
@@ -16,14 +16,10 @@
 
 import attr
```
This file is one we'll definitely want to discuss a bit more, especially as it relates to changing turbine values of specific turbines in a farm, so not necessarily a change in the reference properties, but I think having exclusively one definition would break compatibility with `FlorisInterface.reinitialize_flow_field`. Originally, I could have gone either way, but the examples refactoring has made me realize that wouldn't work because the farm values would break the chain here. Basically the way it was previously was similar enough to how v2 worked to allow for similar workflows, but now we'd be having to wait until there is complete multi-turbine support as opposed to setting ourselves up for it and enabling v2 methods support.
 
 This is a bit long winded, but I think it's worth discussing in more detail, and we can try to find a better way to map the `Ct` and `Cp` functionality to be properly vectorized.

RHammond2 commented on src/floris/simulation/floris.py:
```diff
@@ -35,44 +31,91 @@
+            n_wind_speeds=self.flow_field.n_wind_speeds,
+            layout_x=input_dict["farm"]["layout_x"],
+            layout_y=input_dict["farm"]["layout_y"],
+        )
```
I will say that I prefer not mapping the turbine dictionary directly to the class and keeping it in farm, that said we could always move the turbine mapping into the farm section.

RHammond2 replied:
Just making a note about the use of turbine_id and how we map turbines so it's not forgotten.

RHammond2 commented on src/floris/simulation/floris.py:
```diff
@@ -35,44 +31,91 @@
+
+        # <<interface>>
+        # Initialize grid and field quanitities
+        self.grid = TurbineGrid(
```
There it is!!

RHammond2 commented on src/floris/simulation/floris.py:
```diff
@@ -103,13 +146,17 @@ def from_yaml(cls, input_file_path: str | Path) -> Floris:
 
     def _prepare_for_save(self) -> dict:
+        logging = {
+            "console": {"enable": True, "level": self.logger.level},
+            "file": {"enable": False, "level": 1},
```
Should these be the inputs as opposed to a hard-coded value in case they're not the provided values?

RHammond2 commented on src/floris/simulation/grid.py:
```diff
@@ -173,33 +155,8 @@ def set_grid(self) -> None:
-            + y_center_of_rotation
-        )
+        # These are the rotated coordinates of the wind turbines based on the wind direction
+        x, y, z = rotate_coordinates_rel_west(self.wind_directions, self.turbine_coordinates_array)
```
I like this being a specifically named function that can be reused as needed and is descriptively named!

RHammond2 commented on src/floris/simulation/grid.py:
```diff
@@ -293,19 +239,22 @@ def set_grid(self) -> None:
-        xmax = max(self.turbine_coordinates_array[:, 0]) + 10 * self.reference_turbine_diameter
-        ymin = min(self.turbine_coordinates_array[:, 1]) - 2 * self.reference_turbine_diameter
-        ymax = max(self.turbine_coordinates_array[:, 1]) + 2 * self.reference_turbine_diameter
+        xmin = min(x) - 2 * self.reference_turbine_diameter
```
The whole next few chunks is so much cleaner now, I like it!

RHammond2 commented on src/floris/simulation/solver.py:
```diff
@@ -25,18 +28,11 @@ def sequential_solver(farm: Farm, flow_field: FlowField, grid: TurbineGrid, mode
-            1
-        )
-    )
+    reference_rotor_diameter = farm.reference_turbine_diameter
```
Just making a note here because it relates to the farm.

RHammond2 commented on src/floris/simulation/solver.py:
```diff
@@ -48,27 +44,34 @@ def sequential_solver(farm: Farm, flow_field: FlowField, grid: TurbineGrid, mode
         z_i = z_i[:, :, :, None, None]
 
+        # NOTE: exponential
+        # start = time.time()
```
FYI, the `time.perf_counter()` method is more accurate and `time.time()` is no longer the recommend approach (and is used in the same way: `elapsed = end - start`)

rafmudaf replied:
Ah good to know! I'll update those in my profiling code.

RHammond2 commented on src/floris/simulation/solver.py:
```diff
@@ -48,27 +44,34 @@ def sequential_solver(farm: Farm, flow_field: FlowField, grid: TurbineGrid, mode
         ct_i = Ct(
             velocities=u,
-            yaw_angle=farm.farm_controller.yaw_angles,
+            yaw_angle=farm.yaw_angles,
```
Having the yaw_angles on the farm object probably makes more sense at this point.

RHammond2 commented on src/floris/simulation/solver.py:
```diff
@@ -160,44 +167,33 @@ def full_flow_sequential_solver(farm: Farm, flow_field: FlowField, grid: FlowFie
-
-        z_i = np.mean(turbine_grid.z[:, :, i:i+1], axis=(3, 4))
-        z_i = z_i[:, :, :, None, None]
+        x_i = farm.coordinates[i].x1
+        y_i = farm.coordinates[i].x2
+        z_i = farm.coordinates[i].x3
```
FYI, I added a convenience property a ways back to be able to do: `x_i, y_i, z_i = farm.coordinates[i].elements`, and one exists for `prime_elements` as well.

RHammond2 commented on src/floris/simulation/turbine.py:
```diff
@@ -154,7 +148,7 @@ def power(
         for j in range(n_wind_speeds):
             for k in range(n_turbines):
-                interpolator = power_interp[i, j, k]
+                interpolator = power_interp
                 p[i, j, k] = interpolator(yaw_effective_velocity[i, j, k])
```
Could this just be one line? Also, can we pass in the full 3-D array and get back the values we want? To think into the keeping the Farm object more, it could also be that we don't allow multiple definitions of the interpolations and we simply map the velocities to these functions.

rafmudaf replied:
Good catch, fixed.

RHammond2 commented on src/floris/simulation/turbine.py:
```diff
@@ -183,22 +177,19 @@ def Ct(
         for j in range(n_wind_speeds):
             for k in range(n_turbines):
-                _fCt = fCt[i, j, k]
+                _fCt = fCt
                 thrust_coefficient[i, j, k] = _fCt(average_velocities[i, j, k])
```
I think this could just be one line too.

rafmudaf replied:
Good catch, fixed.

RHammond2 commented on src/floris/simulation/wake_deflection/gauss.py:
```diff
@@ -69,16 +69,16 @@ class GaussVelocityDeflection(BaseModel):
     eps_gain: float = float_attrib(default=0.2)
     use_secondary_steering: bool = bool_attrib(default=True)
 
+    model_string = "gauss"
```
I think for type checking that will get weird, so I'd just do `model_string: str = "gauss"`

rafmudaf replied:
How does type-checking work for a `@property`? Is type checking a static check? As in, does it matter that `model_string` is actually mapping to a `@property` and will the type checking effectively check the input argument to the `@property`?

RHammond2 commented on src/floris/simulation/wake_velocity/curl.py:
```diff
@@ -70,7 +70,7 @@ class CurlVelocityDeficit(BaseClass):
     model_grid_resolution: Union[List[float], Vec3] = attr.ib(
         default=[250, 100, 75],
-        converter=convert_to_Vec3,
+        # converter=convert_to_Vec3,
```
Just curious what the reason for commenting this out was because I thought it was supposed to be a Vec3 object.

rafmudaf replied:
Ah I deleted that function from utilities and commented it here so that the curl model could be imported. However, this is not currently used so I was planning to address this when we reenable curl. Maybe the `convert_to_vec3` function should be moved to the curl model for future reference.

RHammond2 commented on src/floris/tools/floris_interface.py:
```diff
@@ -177,12 +174,12 @@ def create_floris(self) -> None:
             self.floris = Floris.from_dict(self.configuration)
         else:
-            raise TypeError("The Floris `configuration` must of type 'dict', 'str', or 'Path'!")
+            raise TypeError("The Floris `configuration` must of type 'dict', 'str', or 'Path'.")
```
No love for the exclamatory!

rafmudaf replied:
Just being a plain ole engineer over here

rafmudaf replied:
What is emotion anyway.

RHammond2 commented on src/floris/tools/floris_interface.py:
```diff
@@ -207,9 +204,10 @@ def calculate_wake(
-        if yaw_angles is not None:
-            yaw_angles = np.array(yaw_angles)
-            self.floris.farm.farm_controller.set_yaw_angles(yaw_angles)
+        # if yaw_angles is not None:
+        # TODO: reenable
+        #     yaw_angles = np.array(yaw_angles)
+        #     self.floris.farm.farm_controller.set_yaw_angles(yaw_angles)
```
Curious as to why we're not allowing a custom set of yaw angles.

RHammond2 commented on src/floris/tools/floris_interface.py:
```diff
@@ -395,73 +393,40 @@ def get_plane_of_points(
         x, y, _ = coords.T
-        max_diameter = max(self.floris.farm.rotor_diameter)
-        hub_height = self.floris.farm.hub_height[0]
+        max_diameter = self.floris.farm.rotor_diameter
+        hub_height = self.floris.farm.hub_height
```
Another spot I'm pointing out for future reference.

RHammond2 commented on src/floris/tools/floris_interface.py:
```diff
@@ -395,73 +393,40 @@ def get_plane_of_points(
+        hub_height = self.floris.farm.hub_height
 
-        # If x1 and x2 bounds are not provided, use rules of thumb
         if normal_vector == "z":  # Rules of thumb for horizontal plane
             if x1_bounds is None:
                 x1_bounds = (min(x) - 2 * max_diameter, max(x) + 10 * max_diameter)
+
             if x2_bounds is None:
                 x2_bounds = (min(y) - 2 * max_diameter, max(y) + 2 * max_diameter)
+
         if normal_vector == "x":  # Rules of thumb for cut plane plane
             if x1_bounds is None:
                 x1_bounds = (min(y) - 2 * max_diameter, max(y) + 2 * max_diameter)
+
             if x2_bounds is None:
                 x2_bounds = (10, hub_height * 2)
+
         if normal_vector == "y":  # Rules of thumb for cut plane plane
             if x1_bounds is None:
                 x1_bounds = (min(x) - 2 * max_diameter, max(x) + 10 * max_diameter)
+
             if x2_bounds is None:
                 x2_bounds = (10, hub_height * 2)
```
It looks like these are no longer used given the new definitions below, is that correct?

RHammond2 commented on src/floris/tools/floris_interface.py:
```diff
@@ -611,13 +572,13 @@ def get_hor_plane(
         df = self.get_plane_of_points(
-            x1_resolution=x_resolution,
-            x2_resolution=y_resolution,
+            # x1_resolution=x_resolution,
+            # x2_resolution=y_resolution,
```
So is commenting all these references out enabling us to run the examples all the way through now?

rafmudaf replied:
Yes, the user-defined resolution was used to recalculate the flow field solution in v2. For v3, we aren't doing that recalculate step, so we get the resolution from the grid directly.

RHammond2 commented on src/floris/tools/visualization.py:
```diff
@@ -23,7 +23,7 @@
 
 
-def plot_turbines(ax, layout_x, layout_y, yaw_angles, D, color=None, wind_direction=270.0):
+def plot_turbines(ax, layout_x, layout_y, yaw_angles, rotor_diameters, color=None, wind_direction=270.0):
```
That makes a lot of sense! (I know you hate exclamations, but I've got love for them)

rafmudaf replied:
There was a comment later in the code about "grabbing diameter" except it was D and so obviously the remnants of my 16 year old self took over. So I thought we ought to avoid a future awkward situation and start calling diameter by its right name.

RHammond2 commented on src/floris/utilities.py:
```diff
@@ -229,7 +255,7 @@ def from_dict(cls, data: dict):
         kwargs = {a.name: data[a.name] for a in cls.__attrs_attrs__ if a.name in data and a.init}
 
         # Map the inputs must be provided: 1) must be initialized, 2) no default value defined
-        required_inputs = [a.name for a in cls.__attrs_attrs__ if a.init and isinstance(a.default, attr._make._Nothing)]
```
We'll want whatever my change was in this spot, but the tests will let us know what works and what doesn't.

It'll be this one: `required_inputs = [a.name for a in cls.__attrs_attrs__ if a.init and isinstance(a.default, attr._make._Nothing)]`

rafmudaf replied:
Oh I think I took the one from the change you had made in my branch. Mine and yours were slightly different so yes lets double check that.

RHammond2 replied:
Ya, the code snippet here is the one that is correct, and it should be reflected in the tests too. I think you had merged the PR before I caught it on my end and realized I had to do a better check.

RHammond2 reviewed (commented):
Looks like we're going to have a lot to talk about!

rafmudaf committed:
Remove redundant variables

RHammond2 commented:
It seems like we took care of everything conceptually in our discussion, but I'm putting the notes here so that it all gets captured.

- `Farm`
  - [ ] `turbine_id` will be a string input that maps to a turbine definition in the "turbine_library" field of the input file, which will be a reference to a separate file in a future update to FLORIS
  - [ ] There will be one turbine definition allowed in the farm and no ability to ad-hoc update parameters on individual turbines until we get to a multi-turbine FLORIS
  - [x] `set_yaw_angles()`with the few live updates from me will be moved to the Farm class
    - [x] FarmController will go entirely
   - Should the class also have attributes called `rotor_diameter_array` for fields that will need to be used in the solvers?
    - ```python
       def __attrs_post_init__(self) -> None:
           ... # the code currently there
           # Define the few needed as the following so they don't need to be recreated multiple times throughout the code
           # Alternatively, this could be a property
           self.rotor_diameter_array = np.resize([self.rotor_diameter], (self.wind_directions.size, self.wind_speeds.size, self.n_turbines))
      ```
  - The `layout_x` and `layout_y` validators, and array generation will be turned back on for multi-turbine

- Higher Level
  -  [ ] `floris.solve_for_viz()` will be updated to have two different functions so that only the required steps are computed for the differing needs
  - [ ] Update the input files
  - [ ] map the logging settings provided to the logging output
  - The change_turbine examples are going to be changed to swap two turbines in a turbine library with the current implementation pared down to its bare bones to be rebuilt later as needed. (not in this PR though)
  - The yaw angle resetting will need to be redesigned to get the math to work out (let me know if you need help there)
  - Get rid of a lot of examples! (not this PR)

Anything else I've missed

RHammond2 committed:
Merge branch 'enhancement/wind_rose' into enhancement/wind_rose

rafmudaf committed:
Support multiple turbine models but one in use

This allows to define multiple turbines in the input file, but it does not enable multi-turbine support in the farm definition.
